common/src/main/java/com/radixdlt/sbor/codec/StructCodec.java

```diff
@@ -417,4 +423,57 @@ static <T, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12> StructCodec<T> fro
             creator, field1, field2, field3, field4, field5, field6, field7, field8, field9,
             field10, field11, field12));
   }
+
+  // Convenience methods for using Variant 2 (Fields) with Records.
+
+  static <R extends Record> StructCodec<R> fromRecordComponents(
+      Class<R> recordClass, CodecMap.CodecResolver codecs) {
+    return fromRecordComponents(TypeToken.of(recordClass), codecs);
+  }
+
+  @SuppressWarnings("unchecked")
+  static <R extends Record> StructCodec<R> fromRecordComponents(
+      TypeToken<R> recordType, CodecMap.CodecResolver codecs) {
+    RecordComponent[] components = recordType.getRawType().getRecordComponents();
+    List<? extends Field<R, ?>> fields =
+        Stream.of(components).map(component -> toField(recordType, codecs, component)).toList();
+    Constructor<? super R> recordConstructor;
+    try {
+      recordConstructor =
+          recordType
+              .getRawType()
+              .getConstructor(
+                  Stream.of(components).map(RecordComponent::getType).toArray(Class<?>[]::new));
+    } catch (NoSuchMethodException nsme) {
+      throw new IllegalStateException(
+          "guaranteed canonical Record constructor not found on %s".formatted(recordType), nsme);
+    }
+    Functions.Func1<List<?>, R> constructor =
+        argumentList -> {
+          try {
+            return (R) recordConstructor.newInstance(argumentList.toArray());
+          } catch (ReflectiveOperationException roe) {
+            throw new IllegalStateException(
+                "failed to invoke %s with %s".formatted(recordConstructor, argumentList), roe);
+          }
+        };
+    return of(Fields.arbitrary(constructor, (List<Field<R, ?>>) fields));
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <R, C> Field<R, C> toField(
+      TypeToken<R> recordType, CodecMap.CodecResolver codecs, RecordComponent component) {
+    Function<R, C> componentGetter =
+        record -> {
+          try {
+            return (C) component.getAccessor().invoke(record);
+          } catch (ReflectiveOperationException roe) {
+            throw new IllegalStateException(
+                "failed to invoke %s on %s".formatted(component.getAccessor(), record), roe);
+          }
+        };
+    TypeToken<C> componentType = (TypeToken<C>) recordType.resolveType(component.getGenericType());
```

Comment:
Ohhhh psych. This is super cool :D - didn't realise that resolve type could do this - it seems to be able to create a TypeToken for the child out of the parent! ✨. Sweet!

Reply:
This works for nested generic types, right?

EG if we had a `SomeFieldType<T>`, and we were encoding `Record<Integer>` where `Record<T>` has a field of `List<SomeFieldType<T>>` then would this work / capture the generic properly as `List<SomeFieldType<Integer>>`? I imagine it would?

Reply:
I assume you talk about situation where `Record<Integer>` is a field in some other record/class.

Then: yes, all these types will be captured correctly. I mean: they are resolvable, and `.resolveType()` is meant to do it, so if I don't have any bug in this PR, it will work as you expect.

(however, if you meant that `Record<Integer>` is just some runtime type of some var, and you serialize that instance - then no, the type won't be captured. A type param will be captured instead, which will result in an invocation like `codecs.of(new TypeToken<T>() {})`, which will fail. In short: the `.resolveType()` cannot win with the type erasure)

Reply:
the tests currently do
`record SimpleRecord(int first, String second, Either<Long, String> third, Option<Boolean> fourth)`;
this kinda demonstrates the `.resolveType()`'s capabilities already, but AFAIU you would like to change it to e.g.:
`record SimpleRecord(int first, String second, Either<Option<Long>, String> third)`;
for the extra challenge, right?

Reply:
I guess what I'm asking is - can this work with registering `record MyRecord<T1, T2>(T, List<T2>)` via `codecMap.registerForGeneric(MyRecord.class, (codecs, concreteTypeToken) -> StructCodec.fromRecordComponents(concreteTypeToken, codecs));`? 👍 

I think the answer is "yes, it totally can :D".

On a side note, I tend to include contextual arguments (such as `codecs`) first in argument lists, and I see you've put `codecs` last here - I don't mind either way, but maybe consistency might be good 👍

Reply:
ahh yes, it should just work fine with current `.registerForGeneric()` 👍 

pulling the helpers up to `codecMap` level? maaaybe, although you always gotta stop somewhere (for "the right abstraction" reasons), and I think maybe `codecMap` does not need to know stuff about possible impls of codecs.

I will try to remember about "contextual arguments first" 👍 (and migrate the current code next time I touch it).